byron/ledger/executable-spec/test/Ledger/Update/Properties.hs

```diff
@@ -355,42 +355,88 @@ instance Embed UPIREG UBLOCK where
 instance Embed UPIVOTES UBLOCK where
   wrapFailed = UPIVOTESFailure
 
-instance HasTrace UBLOCK where
-  initEnvGen = Update.upiEnvGen
+instance Embed UPIEND UBLOCK where
+  wrapFailed = UPIENDFailure
 
-  sigGen _env UBlockState {upienv, upistate} =
-    Gen.frequency [ (4, generateOnlyVotes)
-                  , (1, generateUpdateProposalAndVotes)
-                  ]
+instance HasTrace UBLOCK where
+  initEnvGen =
+    do
+      let numberOfGenesisKeys = 7
+      dms <- Update.dmapGen numberOfGenesisKeys
+      -- We don't want a large value of @k@, otherwise we won't see many confirmed proposals or
+      -- epoch changes. The problem here is that the initial environment does not know anything
+      -- about the trace size, and @k@ should be a function of it.
+      pure (Slot 0, dms, BlockCount 10, numberOfGenesisKeys)
+
+  sigGen _env UBlockState {upienv, upistate} = do
+    let rpus = Update.registeredProtocolUpdateProposals upistate
+    (anOptionalUpdateProposal, aListOfVotes) <-
+      -- We want to generate update proposals when there is none registered. Otherwise we won't get
+      -- any proposals or votes.
+      if Set.null (dom rpus)
+      then generateUpdateProposalAndVotes
+      else Gen.frequency [ (5, generateOnlyVotes)
+                         , (1, generateUpdateProposalAndVotes)
+                         ]
+    -- Don't shrink the issuer as this won't give us additional insight on a test failure.
+    aBlockIssuer <- Gen.prune $
+      -- Pick a delegate from the delegation map
+      Gen.element $ Bimap.elems (Update.delegationMap upienv)
+
+    let
+      -- Generate a list of protocol version endorsements. For this we look at the current
+      -- endorsements, and confirmed and stable proposals.
+      --
+      -- If there are no endorsements, then the confirmed and stable proposals provide fresh
+      -- protocol versions that can be endorsed.
+      endorsementsList :: [(ProtVer, Set Core.VKeyGenesis)]
+      endorsementsList = endorsementsMap `Map.union` emptyEndorsements
+                       & Map.toList
+        where
+          emptyEndorsements :: Map ProtVer (Set Core.VKeyGenesis)
+          emptyEndorsements = zip stableAndConfirmedVersions (repeat Set.empty)
+                            & Map.fromList
+            where
```

Comment:
While I get that you want to reduce the scope of these to only the function they're used in, I think three layers of `where` is a bit much! It's probably safe to have these in one layer of `where`, or even all in the `let`.

Reply:
Give me a `keyword` and I will use it ad-absurdum :stuck_out_tongue: 

Good point, but I'm not so sure about this one. I do like the fact that the structure gives the reader more information about where function is used. I guess this would read nicer in a text editor that allows to fold the `where` clauses. 

On the other hand such deep nesting might indicate a problem with the way I broke down the function :man_shrugging: